docs/docs/dev_docs/testing/cheat_codes.md

```diff
@@ -0,0 +1,449 @@
+///     uint256 private leet = 1337; // slot 0
+///     mapping(address => uint256) public balances; // base slot 1
+/// }
+
+// find the storage slot for key `0xdead` in the balance map.
+const address = BigInt('0x000000000000000000000000000000000000dead');
+const slot = cc.l1.keccak256(BigInt(1), key) 
+// store balance of 0xdead as 100
+await cc.l1.store(contractAddress, slot, 100n);
+```
+
+### startPrank
+
+#### Function Signature
+```js 
+public async startPrank(who: EthAddress): Promise<void>
+```
+
+#### Description  
+Start impersonating an L1 account
+Sets msg.sender for all subsequent calls until stopPrank is called.
+
+#### Example
+```js
+await cc.l1.startPrank(EthAddress.fromString(address));
+```
+
+### stopPrank
+
+#### Function Signature
+```js
+public async stopPrank(who: EthAddress): Promise<void>
+```
+
+#### Description
+Stop impersonating an L1 account  
+Stops an active prank started by startPrank, resetting msg.sender to the values before startPrank was called.
+
+#### Example
+```js  
+await cc.l1.stopPrank(EthAddress.fromString(address)) 
+```
+
+### getBytecode
+
+#### Function Signature
+```js
+public async getBytecode(contract: EthAddress): Promise<`0x${string}`>
+```
+
+#### Description  
+Get the bytecode for an L1 contract
+
+#### Example
+```js
+const bytecode = await cc.l1.getBytecode(contract) // 0x6080604052348015610010...
+```
+
+### etch
+
+#### Function Signature 
+```js
+public async etch(contract: EthAddress, bytecode: `0x${string}`): Promise<void> 
+```
+
+#### Description
+Set the bytecode for an L1 contract
+
+#### Example  
+```js
+const bytecode = `0x6080604052348015610010...`
+await cc.l1.etch(contract, bytecode)
+console.log(await cc.l1.getBytecode(contract)) // 0x6080604052348015610010...
+```
+
+## L2 related cheatcodes
+These are cheatcodes specific to manipulating the state of Aztec rollup.
+
+### Interface
+```
+// Get the current L2 block number
+public async blockNumber(): Promise<number>
+
+// Set time of the next execution on L2. It also modifies time on L1 for next execution and stores this time as the last rollup block on the rollup contract.
+public async warp(to: number): Promise<void>
+
+// Loads the value stored at the given slot in the public storage of the given contract.
+public async loadPublic(who: AztecAddress, slot: Fr | bigint): Promise<Fr>
+
+// Computes the slot value for a given map and key.
+public computeSlotInMap(baseSlot: Fr | bigint, key: Fr | bigint): Fr
+```
+
+### blockNumber
+
+#### Function Signature
+```js
+public async blockNumber(): Promise<number>
+```
+
+#### Description  
+Get the current L2 block number
+
+#### Example
+```js
+const blockNumber = await cc.l2.blockNumber()
+```
+
+### warp 
+
+#### Function Signature
+```js
+public async warp(to: number): Promise<void>
+```
+
+#### Description
+Set time of the next execution on L2 and L1 for next execution.
+Like with the corresponding L1 cheatcode, time can only be set in the future, not the past.
```

Comment:
Will the cheatcode throw an error if called in the past, or will things simply break? If they are just breaking, better to throw explicitly.